File: src/EditorFeatures/VisualBasicTest/Recommendations/Queries/IntoKeywordRecommenderTests.vb
' Licensed to the .NET Foundation under one or more agreements.
' The .NET Foundation licenses this file to you under the MIT license.
' See the LICENSE file in the project root for more information.

Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.UnitTests.Recommendations.Queries
    <Trait(Traits.Feature, Traits.Features.KeywordRecommending)>
    Public Class IntoKeywordRecommenderTests
        Inherits RecommenderTests

        <Fact, WorkItem("http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/543191")>
        Public Sub IntoAfterAnonymousObjectCreationExprTest()
            Dim method = <MethodBody>
                            Dim q1 = From num In New Integer() {4, 5} Group By i1 = New With {.Key = num} |
                         </MethodBody>

            VerifyRecommendationsAreExactly(method, "Into")
        End Sub

        <Fact, WorkItem("http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/543193")>
        Public Sub IntoAfterExprRangeVariableInGroupByTest()
            Dim method = <MethodBody>
                            Dim q1 = From num In New Integer() {4, 5} Group By num |
                         </MethodBody>

            VerifyRecommendationsAreExactly(method, "Into")
        End Sub

        <Fact, WorkItem("http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/543214")>
        Public Sub IntoImmediatelyAfterAnonymousObjectCreationExprTest()
            Dim method = <MethodBody>
                            Dim q1 = From num In New Integer() {4, 5} Group By i1 = New With {.Key = num}|
                         </MethodBody>

            VerifyRecommendationsAreExactly(method, "Into")
        End Sub

        <Fact, WorkItem("http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/543232")>
        Public Sub IntoAfterNestedAggregateFromClauseTest()
            VerifyRecommendationsContain(<MethodBody>Dim q1 = Aggregate i1 In arr From i4 In arr |</MethodBody>, "Into")
        End Sub
    End Class
End Namespace

File: src/EditorFeatures/CSharpTest2/Recommendations/IntoKeywordRecommenderTests.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System.Threading.Tasks;
using Microsoft.CodeAnalysis.Test.Utilities;
using Xunit;

namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Recommendations;

[Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
public sealed class IntoKeywordRecommenderTests : KeywordRecommenderTests
{
    [Fact]
    public Task TestNotAtRoot_Interactive()
        => VerifyAbsenceAsync(SourceCodeKind.Script,
@"$$");

    [Fact]
    public Task TestNotAfterClass_Interactive()
        => VerifyAbsenceAsync(SourceCodeKind.Script,
            """
            class C { }
            $$
            """);

    [Fact]
    public Task TestNotAfterGlobalStatement_Interactive()
        => VerifyAbsenceAsync(SourceCodeKind.Script,
            """
            System.Console.WriteLine();
            $$
            """);

    [Fact]
    public Task TestNotAfterGlobalVariableDeclaration_Interactive()
        => VerifyAbsenceAsync(SourceCodeKind.Script,
            """
            int i = 0;
            $$
            """);

    [Fact]
    public Task TestNotInUsingAlias()
        => VerifyAbsenceAsync(
@"using Goo = $$");

    [Fact]
    public Task TestNotInGlobalUsingAlias()
        => VerifyAbsenceAsync(
@"global using Goo = $$");

    [Fact]
    public Task TestNotInEmptyStatement()
        => VerifyAbsenceAsync(AddInsideMethod(
@"$$"));
    [Fact]
    public Task TestNotInSelectMemberExpressionOnlyADot()
        => VerifyAbsenceAsync(AddInsideMethod(
@"var y = from x in new [] { 1,2,3 } select x.$$"));
    [Fact]
    public Task TestNotInSelectMemberExpression()
        => VerifyAbsenceAsync(AddInsideMethod(
@"var y = from x in new [] { 1,2,3 } select x.i$$"));
    [Fact]
    public Task TestAfterJoinRightExpr()
        => VerifyKeywordAsync(AddInsideMethod(
            """
            var q = from x in y
                      join a in e on o1 equals o2 $$
            """));

    [Fact]
    public Task TestAfterJoinRightExpr_NotAfterInto()
        => VerifyAbsenceAsync(AddInsideMethod(
            """
            var q = from x in y
                      join a.b c in o1 equals o2 into $$
            """));

    [Fact]
    public Task TestNotAfterEquals()
        => VerifyAbsenceAsync(AddInsideMethod(
            """
            var q = from x in y
                      join a.b c in o1 equals $$
            """));

    [Fact]
    public Task TestAfterSelectClause()
        => VerifyKeywordAsync(AddInsideMethod(
            """
            var q = from x in y
                      select z
                      $$
            """));

    [Fact]
    public Task TestAfterSelectClauseWithMemberExpression()
        => VerifyKeywordAsync(AddInsideMethod(
            """
            var q = from x in y
                      select z.i
                      $$
            """));

    [Fact]
    public Task TestAfterSelectClause_NotAfterInto()
        => VerifyAbsenceAsync(AddInsideMethod(
            """
            var q = from x in y
                      select z
                      into $$
            """));

    [Fact]
    public Task TestAfterGroupClause()
        => VerifyKeywordAsync(AddInsideMethod(
            """
            var q = from x in y
                      group z by w
                      $$
            """));

    [Fact]
    public Task TestAfterGroupClause_NotAfterInto()
        => VerifyAbsenceAsync(AddInsideMethod(
            """
            var q = from x in y
                      group z by w
                      into $$
            """));

    [Fact]
    public Task TestNotAfterSelect()
        => VerifyAbsenceAsync(AddInsideMethod(
            """
            var q = from x in y
                      select $$
            """));

    [Fact]
    public Task TestNotAfterGroupKeyword()
        => VerifyAbsenceAsync(AddInsideMethod(
            """
            var q = from x in y
                      group $$
            """));

    [Fact]
    public Task TestNotAfterGroupExpression()
        => VerifyAbsenceAsync(AddInsideMethod(
            """
            var q = from x in y
                      group x $$
            """));

    [Fact]
    public Task TestNotAfterGroupBy()
        => VerifyAbsenceAsync(AddInsideMethod(
            """
            var q = from x in y
                      group x by $$
            """));
}
